android/greengrass-nucleus-android/src/main/java/com/aws/greengrass/android/managers/AndroidBasePackageManager.java

```diff
@@ -0,0 +1,578 @@
+        PackageInfo packageInfo = getApkPackageInfo(apkPath);
+
+        return createAndroidPackageIdentifier(packageInfo);
+    }
+
+    /**
+     * Create AndroidPackageIdentifier from Android PackageInfo.
+     *
+     * @param packageInfo Android PackageInfo
+     */
+    private AndroidPackageIdentifier createAndroidPackageIdentifier(PackageInfo packageInfo) {
+        return new AndroidPackageIdentifier(packageInfo.packageName,
+                new Semver(packageInfo.versionName), getVersionCode(packageInfo));
+    }
+
+    /**
+     * Get APK installer callable.
+     *
+     * @param cmdLine #install_package command line
+     * @param packageName APK should contains that package
+     * @param logger logger to log to
+     * @return Callable Object to call installAPK()
+     * @throws IOException      on errors
+     */
+    @Override
+    public AndroidCallable getApkInstaller(String cmdLine, String packageName,
+                                             @Nullable Logger logger) throws IOException {
+        if (Utils.isEmpty(cmdLine)) {
+            throw new IOException("Expected " + APK_INSTALL_CMD + " command but got empty line");
+        }
+
+        String[] cmdParts = cmdLine.split("\\s+");
+        if (cmdParts.length < 2 || cmdParts.length > 3) {
+            throw new IOException("Invalid " + APK_INSTALL_CMD + " command line, expected " + APK_INSTALL_CMD_EXAMPLE);
+        }
+        String cmd = cmdParts[0];
+        if (!APK_INSTALL_CMD.equals(cmd)) {
+            throw new IOException("Unexpected command, expecting " + APK_INSTALL_CMD_EXAMPLE);
+        }
+
+        String apkFile = cmdParts[1];
+        boolean force = false;
+        if (cmdParts.length == 3) {
+            String forceString = cmdParts[2];
+            if (forceString.startsWith(APK_INSTALL_FORCE)) {
+                if (APK_INSTALL_FORCE.equals(forceString)) {
+                    force = true;
+                } else {
+                    String[] forceParts = cmdParts[2].split("=");
+                    if (forceParts.length != 2) {
+                        throw new IOException("Unexpected force part of command, expecting " + APK_INSTALL_CMD_EXAMPLE);
+                    }
+                    force = Coerce.toBoolean(forceParts[1]);
+                }
+            } else {
+                throw new IOException("Missing force part of command, expecting " + APK_INSTALL_CMD_EXAMPLE);
+            }
+        }
+        return new Installer(apkFile, packageName, force, logger);
+    }
+
+    /**
+     * Install APK file.
+     *
+     * @param apkPath   path to APK file
+     * @param packageName APK should contains that package
+     * @param logger optional logger to log to
+     * @throws IOException      on errors
+     * @throws InterruptedException when thread has been interrupted
+     */
+    @Override
+    public void installAPK(@NonNull String apkPath, @NonNull String packageName, boolean force,
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 14. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 283-300 into a separate method.